### PR TITLE
Cache Auth token AB#12058

### PR DESCRIPTION
**Pull request by sslaws**
# Fixes or Implements [AB#12058](https://dev.azure.com/qslvic/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/12058), [AB#12067](https://dev.azure.com/qslvic/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/12067)

## Description

Adds the ability  to cache the auth token when using AuthenticateAsUser in the common AuthenticationDelegate.
Immunization was updated to make use of the cache and remove the local service cache.
Medication was updated to make use of the new cache (Salesforce).

If you are reviewing this PR, please adhere to the guidelines as [documented](https://github.com/bcgov/healthgateway/wiki/prguidance).

## Testing

-   [X] Unit Tests Updated
-   [ ] Functional Tests Updated
-   [ ] Not Required

### UI Changes

No

### Browsers Tested

-   [X] Chrome
-   [ ] Safari
-   [ ] Edge
-   [ ] Firefox

Provide an explanation for any test(s) not completed.

## Notes/Additional Comments

Any additional notes or comments that may be relevant. Include any specialized deployment steps here.

## Items to Review:

-   [General PR Guidelines](https://github.com/bcgov/healthgateway/wiki/PRguidance)

-   Fulfills Work Item Requirements

    -   The changes meet/implement story's acceptance criteria. Fixes identified problems if bug.
    -   Changes not directly related to the task requirements need to be documented on the PR.

-   Compilation / Tests

    -   The changes do not break compilation and/or unit or functional tests; unless other item addresses them specifically.

-   Logic Problems / Functional Behaviour

    -   The changes work as intended and do not introduce problems.

-   Performance

    -   The changes do not introduce obvious performance issues.

-   Documented Standards

    -   The changes adhere to the team's documented [Coding Standards](https://github.com/bcgov/healthgateway/wiki/standards).

-   Readability / Maintainability
    -   The changes can be easily understood/read and allow for future enhancements without major refactoring. Readability is preferred over "clever code".
    -   Reasoning for changes needs to be clearly articulated. Disagreements should be arbitrated by a third developer.

### CODE REVIEW

**Review by pagustinqt20 — DISMISSED** [review]

File: Apps/Common/src/AccessManagement/Authentication/AuthenticationDelegate.cs [hunk 1]
```diff
@@ -23,57 +23,114 @@ namespace HealthGateway.Common.AccessManagement.Authentication
...
         {
-            this.logger.LogDebug($"Authenticating Direct Grant as User: {tokenRequest.Username}");
-            Task<IAuthModel> authenticating = this.ResourceOwnerPasswordGrant(tokenUri, tokenRequest);
+            (JWTModel jwtModel, _) = this.AuthenticateUser(tokenUri, tokenRequest, cacheEnabled);
+            return jwtModel;
+        }
 
-            JWTModel jwtModel = (authenticating.Result as JWTModel)!;
-            this.logger.LogDebug($"Finished authenticating User: {tokenRequest.Username}");
+            /// <inheritdoc/>
```
Comment (MikeLyttle) [inline Apps/Common/src/AccessManagement/Authentication/AuthenticationDelegate.cs:84]: Nit: weird indentation on this comment.

File: Apps/Common/src/AccessManagement/Authentication/AuthenticationDelegate.cs [hunk 2]
```diff
@@ -126,12 +183,14 @@ private async Task<IAuthModel> ResourceOwnerPasswordGrant(Uri tokenUri, ClientCr
                 content.Headers.Clear();
                 content.Headers.Add(@"Content-Type", @"application/x-www-form-urlencoded");
 
-                using HttpResponseMessage response = await client.PostAsync(tokenUri, content).ConfigureAwait(true);
-
-                string jwtTokenResponse = await response.Content.ReadAsStringAsync().ConfigureAwait(true);
-                this.logger.LogTrace($"JWT Token response: {jwtTokenResponse}");
-                response.EnsureSuccessStatusCode();
-                authModel = JsonSerializer.Deserialize<JWTModel>(jwtTokenResponse)!;
+                using (Task<HttpResponseMessage> task = client.PostAsync(tokenUri, content))
```
Comment (MikeLyttle) [inline Apps/Common/src/AccessManagement/Authentication/AuthenticationDelegate.cs:186]: Nit: the new `using` syntax introduced in C# 8.0 is preferred.  There is also an indentation issue with the block currently.

**Review by MikeLyttle — APPROVED** [review]

File: Apps/Common/src/AccessManagement/Authentication/AuthenticationDelegate.cs (hunk 1 above)
Comment (sslaws) [reply Apps/Common/src/AccessManagement/Authentication/AuthenticationDelegate.cs]: Fixed

File: Apps/Common/src/AccessManagement/Authentication/AuthenticationDelegate.cs (hunk 2 above)
Comment (sslaws) [reply Apps/Common/src/AccessManagement/Authentication/AuthenticationDelegate.cs:186]: Oddly enough I was getting a disposed error so I switched the style and it worked.  Fixed formatting

**Review by BrianMaki — APPROVED** [review]

**Review by pagustinqt20 — APPROVED** [review]

**Pull request merged**